### PR TITLE
Use builder pattern when constructing widgets

### PR DESCRIPTION
**Pull request by peterstuart**
This adds a macro that makes it possible to `#[derive(Builder)]` on structs.

## Testing

Everything should work the same as before.

### CODE REVIEW

File: cherry_macros/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            });
+
+        if !omit {
+            functions.extend(quote! {
+                pub fn #name(mut self, value: #ty) -> Self {
+                    self.#name = value;
+                    self
+                }
+            });
+        }
```
Comment (peterstuart) [inline cherry_macros/src/lib.rs]: For each non-omitted field in the struct, generate a function with the same name that sets that field and returns `self`.

File: examples/container.rs [hunk 2]
```diff
@@ -18,78 +18,56 @@ fn main() -> Result<(), Infallible> {
...
-        children: vec![
+        }))
+        .corner_radii(Some(CornerRadii::new(Size::new(10, 10))))
+        .justification(Justification::SpaceBetween)
+        .padding(Insets::all(10))
+        .width(Some(200))
+        .children(vec![
             colored_container(Rgb888::RED, 25).boxed(),
             colored_container(Rgb888::GREEN, 40).boxed(),
             colored_container(Rgb888::BLUE, 55).boxed(),
```
Comment (peterstuart) [inline examples/container.rs:38]: Chaining like this seems much nicer than the `Options` structs I was using before.

File: cherry_macros/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    for field in fields {
+        if should_omit(&field) {
+            continue;
+        }
+
+        let name = &field.ident.as_ref().unwrap();
+        let ty = &field.ty;
+
+        let function = match option_inner_type(ty) {
+            Some(option_type) => quote! {
```
Comment (peterstuart) [inline cherry_macros/src/lib.rs:36]: If the type of the field is `Option<Something>`, get the inner type

File: cherry_macros/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                pub fn #name(mut self, value: #option_type) -> Self {
```
Comment (peterstuart) [inline cherry_macros/src/lib.rs:37]: Don't require the `Some` wrapper when calling the setter.

**Review by helixbass — APPROVED** [review]
@peterstuart nice, I was able to follow the procedural macro code pretty well. And definitely nice to be able to use this pattern for these types of widget structs. When I was reading about procedural macros I saw that in [this tutorial/"workshop"](https://github.com/dtolnay/proc-macro-workshop) he actually uses building a `derive(Builder)` as an example. And it looks like [`derive_builder`](https://crates.io/crates/derive_builder) is maybe the most widely used crate for this type of thing, I wonder how its API/implementation compare to what you did

**Comment by peterstuart** [comment]
Oh nice, I hadn't seen that tutorial. The reason I didn't go with any of the existing builder macro crates is that they all end up with a separate "Builder" type for the struct, which also pushes some errors to runtime instead of compile-time (the `build` function returns a result). I think that's worse for my specific case since it's more verbose and error-prone.

**Pull request merged**